File: 20-agent-config/23-agent-tools/python/23-agent-tools.py
#!/usr/bin/env python3
"""
23-agent-tools.py — thin HTTP adapter for the tools demo UI.

  1. Serve index.html
  2. JSON bootstrap
  3. Yahoo Finance headlines
  4. Bridge browser SSE → agent_core.generate_stream() (tool loop)

LaunchDarkly work lives in agent_core.py (completion_config + tools + track_tool_call).
"""

from __future__ import annotations

import json
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))

from agent_core import (  # noqa: E402
    PERSONAS,
    config_key,
    generate_stream,
    init_launchdarkly,
    persona_by_id,
)
from yahoo_news import (  # noqa: E402
    DEFAULT_TICKER_1,
    DEFAULT_TICKER_2,
    fetch_stories_for_tickers,
    get_last_pair_cached,
)

APP_BANNER = "23-agent-tools[python]"
PORT = 8230


class Handler(BaseHTTPRequestHandler):
    server_version = "AgentToolsHTTP/1.0"

    def log_message(self, fmt: str, *args) -> None:
        sys.stderr.write("%s - %s\n" % (self.address_string(), fmt % args))

    def do_GET(self) -> None:  # noqa: N802
        parsed = urlparse(self.path)
        path = parsed.path
        qs = parse_qs(parsed.query)

        if path in {"/", "/index.html"}:
            self._serve_file(HERE / "index.html", "text/html; charset=utf-8")
            return

        if path == "/api/bootstrap":
            cached = get_last_pair_cached()
            body = {
                "appBanner": APP_BANNER,
                "personas": [
                    {
                        "id": p.id,
                        "name": p.name,
                        "profile": p.profile,
                        "model": p.model,
                        "anonymous": p.anonymous,
                    }
                    for p in PERSONAS
                ],
                "defaultTickers": {
                    "ticker1": (cached or {}).get("ticker1") or DEFAULT_TICKER_1,
                    "ticker2": (cached or {}).get("ticker2") or DEFAULT_TICKER_2,
                },
                "cachedStories": cached,
                "mode": "launchdarkly",
                "provider": "AgentControl",
                "model": f"config:{config_key()}",
                "configKey": config_key(),
            }
            self._send_json(200, body)
            return

        if path == "/api/stories":
            ticker1 = (qs.get("ticker1") or [DEFAULT_TICKER_1])[0]
            ticker2 = (qs.get("ticker2") or [DEFAULT_TICKER_2])[0]
            try:
                self._send_json(200, fetch_stories_for_tickers(ticker1, ticker2, 2))
            except Exception as exc:  # noqa: BLE001
                self._send_json(500, {"error": str(exc)})
            return

        self.send_error(404)

    def do_POST(self) -> None:  # noqa: N802
        parsed = urlparse(self.path)
        if parsed.path != "/api/generate":
            self.send_error(404)
            return

        length = int(self.headers.get("Content-Length") or 0)
        raw = self.rfile.read(length) if length else b"{}"
        try:
            payload = json.loads(raw.decode("utf-8") or "{}")
        except json.JSONDecodeError:
            self._send_json(400, {"error": "Invalid JSON body."})
            return

        persona_id = str(payload.get("personaId") or PERSONAS[0].id)
        persona = persona_by_id(persona_id) or PERSONAS[0]
        stories = payload.get("stories") if isinstance(payload.get("stories"), list) else []

        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.send_header("Connection", "close")
        self.end_headers()

        try:
            for event in generate_stream(persona, stories):
                line = "data: " + json.dumps(event, default=str) + "\n\n"
                self.wfile.write(line.encode("utf-8"))
                self.wfile.flush()
        except (BrokenPipeError, ConnectionResetError):
            return
        except Exception as exc:  # noqa: BLE001
            try:
                err = "data: " + json.dumps({"type": "error", "message": str(exc)}) + "\n\n"
                self.wfile.write(err.encode("utf-8"))
                self.wfile.flush()
            except Exception:  # noqa: BLE001
                return

    def _serve_file(self, path: Path, content_type: str) -> None:
        if not path.is_file():
            self.send_error(404)
            return
        data = path.read_bytes()
        self.send_response(200)
        self.send_header("Content-Type", content_type)
        self.send_header("Cache-Control", "no-store")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _send_json(self, status: int, body: object) -> None:
        raw = json.dumps(body, default=str).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.send_header("Content-Length", str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)


def main() -> None:
    init_launchdarkly()
    server = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    print(APP_BANNER)
    print(f"Open http://127.0.0.1:{PORT}/")
    print(f"LD_AGENT_CONFIG_KEY={config_key()}")
    print("Press Ctrl+C to stop.")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopped.")


if __name__ == "__main__":
    main()

File: 20-agent-config/23-agent-tools/node/23-agent-tools.js
#!/usr/bin/env node
/**
 * 23-agent-tools.js — thin HTTP adapter for the tools demo UI.
 *
 *   GET /              → index.html
 *   GET /api/bootstrap → personas, tickers, cached stories, config key
 *   GET /api/stories   → Yahoo headlines
 *   POST /api/generate → SSE (tool loop + trackToolCall)
 *
 * LaunchDarkly work lives in agentCore.js (completionConfig + tools + trackToolCall).
 */

"use strict";

const http = require("http");
const fs = require("fs");
const path = require("path");
const { URL } = require("url");

const {
  PERSONAS,
  configKey,
  generateStream,
  initLaunchDarkly,
  personaById,
} = require("./agentCore");
const {
  DEFAULT_TICKER_1,
  DEFAULT_TICKER_2,
  fetchStoriesForTickers,
  getLastPairCached,
} = require("./yahooNews");

const APP_BANNER = "23-agent-tools[node]";
const PORT = Number(process.env.PORT || 8231);
const ROOT = __dirname;

function sendJson(res, status, body) {
  const raw = Buffer.from(JSON.stringify(body), "utf8");
  res.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Content-Length": raw.length,
    "Cache-Control": "no-store",
  });
  res.end(raw);
}

function sendFile(res, filePath, contentType) {
  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
    res.writeHead(404);
    res.end("Not found");
    return;
  }
  const data = fs.readFileSync(filePath);
  res.writeHead(200, {
    "Content-Type": contentType,
    "Content-Length": data.length,
    "Cache-Control": "no-store",
  });
  res.end(data);
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on("data", (chunk) => chunks.push(chunk));
    req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
    req.on("error", reject);
  });
}

async function handleGenerate(req, res) {
  let payload = {};
  try {
    payload = JSON.parse((await readBody(req)) || "{}");
  } catch {
    sendJson(res, 400, { error: "Invalid JSON body." });
    return;
  }

  const personaId = String(payload.personaId || PERSONAS[0].id);
  const persona = personaById(personaId) || PERSONAS[0];
  let stories = payload.stories;
  if (!Array.isArray(stories)) stories = [];

  res.writeHead(200, {
    "Content-Type": "text/event-stream; charset=utf-8",
    "Cache-Control": "no-store",
    Connection: "close",
  });

  try {
    for await (const event of generateStream(persona, stories)) {
      res.write(`data: ${JSON.stringify(event)}\n\n`);
    }
  } catch (exc) {
    if (exc.code !== "EPIPE" && exc.code !== "ECONNRESET") console.error(exc);
  }
  res.end();
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url || "/", `http://${req.headers.host || "127.0.0.1"}`);
  const pathname = url.pathname;

  try {
    if (req.method === "GET" && (pathname === "/" || pathname === "/index.html")) {
      sendFile(res, path.join(ROOT, "index.html"), "text/html; charset=utf-8");
      return;
    }

    if (req.method === "GET" && pathname === "/api/bootstrap") {
      const cached = getLastPairCached();
      sendJson(res, 200, {
        appBanner: APP_BANNER,
        personas: PERSONAS.map((p) => ({
          id: p.id,
          name: p.name,
          profile: p.profile,
          model: p.model,
          anonymous: Boolean(p.anonymous),
        })),
        defaultTickers: {
          ticker1: (cached && cached.ticker1) || DEFAULT_TICKER_1,
          ticker2: (cached && cached.ticker2) || DEFAULT_TICKER_2,
        },
        cachedStories: cached,
        mode: "launchdarkly",
        provider: "AgentControl",
        model: `config:${configKey()}`,
        configKey: configKey(),
      });
      return;
    }

    if (req.method === "GET" && pathname === "/api/stories") {
      const ticker1 = url.searchParams.get("ticker1") || DEFAULT_TICKER_1;
      const ticker2 = url.searchParams.get("ticker2") || DEFAULT_TICKER_2;
      sendJson(res, 200, await fetchStoriesForTickers(ticker1, ticker2, 2));
      return;
    }

    if (req.method === "POST" && pathname === "/api/generate") {
      await handleGenerate(req, res);
      return;
    }

    res.writeHead(404);
    res.end("Not found");
  } catch (exc) {
    console.error(exc);
    if (!res.headersSent) sendJson(res, 500, { error: String(exc.message || exc) });
    else res.end();
  }
});

async function main() {
  await initLaunchDarkly();
  server.listen(PORT, "127.0.0.1", () => {
    console.log(APP_BANNER);
    console.log(`Open http://127.0.0.1:${PORT}/`);
    console.log(`LD_AGENT_CONFIG_KEY=${configKey()}`);
    console.log("Press Ctrl+C to stop.");
  });
}

main().catch((exc) => {
  console.error(exc.message || exc);
  process.exit(1);
});
